File: scripts/custom-sections-diff.py
#!/usr/bin/env python
"""custom-sections-diff.py — the EXTERNAL gate for what wasmrt's assembler writes OUTSIDE the
code: custom sections from `@custom`, the `name` section, and branch hints.

    python scripts/custom-sections-diff.py <wasmrt.exe> <file.wat | @list.txt>...

Each file is assembled twice — by `wasmrt wat` and by `wasm-tools parse` (the `wat` crate that
wasmtime reads text with) — and the two outputs are compared on:

  * the ORDER of sections, customs included (placement is the whole meaning of `(after func)`);
  * every custom section's bytes, exactly;
  * branch hints by (function, offset PAST THE LOCALS VECTOR, value). The raw offsets are not
    comparable: wasm-tools packs `(local i32 i32)` as `01 02 7f`, wasmrt as `02 01 7f 01 7f`,
    and a hint's offset counts from the start of the body. Both are correct for their own body.

Exit status is non-zero if any file differs, so the gate CAN fail — checked by feeding it a
module whose placement was deliberately wrong when it was written.

WHY IT EXISTS: wasmrt's own decoder reads the `name` section leniently and ignores everything
else in a custom section, so no amount of round-tripping through wasmrt could catch a wrong byte
here. Agreement between components that learned a format from each other is not evidence
(`cmem/best-practices.md` §3.8b); only an outside reader is.
"""
import subprocess
import sys
import tempfile
import os
import shutil


def uleb(b, p):
    v = shift = 0
    while True:
        x = b[p]
        p += 1
        v |= (x & 0x7F) << shift
        if not x & 0x80:
            return v, p
        shift += 7


def sections(b):
    out, p = [], 8
    while p < len(b):
        sid = b[p]
        n, q = uleb(b, p + 1)
        out.append((sid, b[q:q + n]))
        p = q + n
    return out


def custom_name(payload):
    n, q = uleb(payload, 0)
    return payload[q:q + n].decode("utf-8", "replace"), payload[q + n:]


def locals_len(code_payload):
    """Length of each body's locals vector, in body order."""
    count, p = uleb(code_payload, 0)
    lens = []
    for _ in range(count):
        size, p = uleb(code_payload, p)
        start = p
        groups, q = uleb(code_payload, p)
        for _ in range(groups):
            _, q = uleb(code_payload, q)
            q = skip_valtype(code_payload, q)
        lens.append(q - start)
        p = start + size
    return lens


def skip_valtype(b, p):
    t = b[p]
    p += 1
    if t in (0x63, 0x64):  # (ref null? ht)
        # heap type: a negative s33 abstract type is one byte; a concrete index is a LEB
        _, p = uleb(b, p)
    return p


def hints(payload, first_def, lens):
    n, p = uleb(payload, 0)
    out = []
    for _ in range(n):
        f, p = uleb(payload, p)
        k, p = uleb(payload, p)
        for _ in range(k):
            off, p = uleb(payload, p)
            size, p = uleb(payload, p)
            val = payload[p:p + size]
            p += size
            out.append((f, off - lens[f - first_def], bytes(val)))
    return out


def import_funcs(secs):
    for sid, pl in secs:
        if sid == 2:
            n, p = uleb(pl, 0)
            funcs = 0
            for _ in range(n):
                for _ in range(2):
                    ln, p = uleb(pl, p)
                    p += ln
                kind = pl[p]
                p += 1
                if kind == 0:
                    funcs += 1
                    _, p = uleb(pl, p)
                else:
                    return None  # other import kinds: not needed by any hint test
            return funcs
    return 0


def describe(b):
    secs = sections(b)
    lens = next((locals_len(pl) for sid, pl in secs if sid == 10), [])
    first = import_funcs(secs) or 0
    layout, customs = [], []
    for sid, pl in secs:
        if sid == 0:
            name, data = custom_name(pl)
            layout.append("custom:" + name)
            if name == "metadata.code.branch_hint":
                customs.append((name, hints(data, first, lens)))
            else:
                customs.append((name, bytes(data)))
        else:
            layout.append(str(sid))
    return layout, customs


def assemble(cmd):
    r = subprocess.run(cmd, capture_output=True)
    return r.returncode == 0, r.stderr.decode("utf-8", "replace").strip()


def main():
    wasmrt, files = os.path.abspath(sys.argv[1]), []
    # `@list.txt` names a file of paths, one per line — a whole corpus overflows a Windows
    # command line.
    for arg in sys.argv[2:]:
        if arg.startswith("@"):
            files += [ln.strip() for ln in open(arg[1:], encoding="utf-8") if ln.strip()]
        else:
            files.append(arg)
    wasm_tools = shutil.which("wasm-tools")
    if not os.path.exists(wasmrt) or not wasm_tools:
        sys.exit(f"need {wasmrt} and wasm-tools on PATH (found: {wasm_tools})")
    bad = 0
    # Agreement that a file is REFUSED compares no bytes at all, so it is counted apart: a probe
    # both tools refuse tested nothing, and the headline must not let it pass for a comparison.
    refused = []
    tmp = tempfile.mkdtemp()
    a, b = os.path.join(tmp, "a.wasm"), os.path.join(tmp, "b.wasm")
    for f in files:
        ok_a, err_a = assemble([wasmrt, "wat", f, "-o", a])
        ok_b, err_b = assemble([wasm_tools, "parse", f, "-o", b])
        if ok_a != ok_b:
            bad += 1
            print(f"DIFF  {f}: wasmrt {'accepts' if ok_a else 'refuses'}, "
                  f"wasm-tools {'accepts' if ok_b else 'refuses'}")
            print(f"        wasmrt:     {err_a[:160]}")
            print(f"        wasm-tools: {err_b[:160]}")
            continue
        if not ok_a:
            refused.append(f)
            continue
        da, db = describe(open(a, "rb").read()), describe(open(b, "rb").read())
        if da != db:
            bad += 1
            print(f"DIFF  {f}")
            if da[0] != db[0]:
                print(f"        layout wasmrt:     {da[0]}")
                print(f"        layout wasm-tools: {db[0]}")
            for (na, xa), (nb, xb) in zip(da[1], db[1]):
                if (na, xa) != (nb, xb):
                    print(f"        custom {na!r}: wasmrt {xa!r:.200}")
                    print(f"        custom {nb!r}: wasm-tools {xb!r:.200}")
    compared = len(files) - bad - len(refused)
    print(f"custom-sections gate: {compared} compared byte-for-byte and agree, "
          f"{len(refused)} refused by both, {bad} differ (of {len(files)})")
    for f in refused:
        print(f"  refused by both: {f}")
    sys.exit(1 if bad else 0)


if __name__ == "__main__":
    main()

File: scripts/custom-sections-diff.ts
#!/usr/bin/env -S deno run --allow-run --allow-read --allow-write --allow-env
// custom-sections-diff.ts — the EXTERNAL gate for what wasmrt's assembler writes OUTSIDE the
// code: custom sections from `@custom`, the `name` section, and branch hints.
//
// SPDX-License-Identifier: MIT OR Apache-2.0
// Copyright (c) 2026 Jon Marcum
//
//   deno run -A scripts/custom-sections-diff.ts <wasmrt.exe> <file.wat | @list.txt>...
//   bun scripts/custom-sections-diff.ts <wasmrt.exe> <file.wat | @list.txt>...
//
// Each file is assembled twice — by `wasmrt wat` and by `wasm-tools parse` (the `wat` crate that
// wasmtime reads text with) — and the two outputs are compared on:
//
//   * the ORDER of sections, customs included (placement is the whole meaning of `(after func)`);
//   * every custom section's bytes, exactly;
//   * branch hints by (function, offset PAST THE LOCALS VECTOR, value). The raw offsets are not
//     comparable: wasm-tools packs `(local i32 i32)` as `01 02 7f`, wasmrt as `02 01 7f 01 7f`,
//     and a hint's offset counts from the start of the body. Both are correct for their own body.
//
// Exit status is non-zero if any file differs, so the gate CAN fail — checked by feeding it an
// assembler wrapper that strips the custom sections, which must be reported as a difference.
//
// WHY IT EXISTS: wasmrt's own decoder reads the `name` section leniently and ignores everything
// else in a custom section, so no amount of round-tripping through wasmrt could catch a wrong byte
// here. Agreement between components that learned a format from each other is not evidence
// (`cmem/best-practices.md` §3.8b); only an outside reader is.
//
// 🔻 Ported from `custom-sections-diff.py` (2026-09-19, the TypeScript scripting rule), function
// for function, and verified by running both over the whole `.wat` corpus for the same verdict.
import { spawnSync } from "node:child_process";
import { existsSync, mkdtempSync, readFileSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join, resolve } from "node:path";
import process from "node:process";

function uleb(b: Uint8Array, p: number): [number, number] {
  let v = 0, shift = 0;
  for (;;) {
    const x = b[p++];
    v |= (x & 0x7f) << shift;
    if (!(x & 0x80)) return [v >>> 0, p];
    shift += 7;
  }
}

type Section = { id: number; payload: Uint8Array };

function sections(b: Uint8Array): Section[] {
  const out: Section[] = [];
  let p = 8;
  while (p < b.length) {
    const id = b[p];
    const [n, q] = uleb(b, p + 1);
    out.push({ id, payload: b.subarray(q, q + n) });
    p = q + n;
  }
  return out;
}

function customName(payload: Uint8Array): [string, Uint8Array] {
  const [n, q] = uleb(payload, 0);
  return [new TextDecoder().decode(payload.subarray(q, q + n)), payload.subarray(q + n)];
}

function skipValtype(b: Uint8Array, p: number): number {
  const t = b[p++];
  // (ref null? ht) — a heap type: an abstract one is a single byte, a concrete index is a LEB.
  if (t === 0x63 || t === 0x64) [, p] = uleb(b, p);
  return p;
}

/** Length of each body's locals vector, in body order. */
function localsLen(codePayload: Uint8Array): number[] {
  let [count, p] = uleb(codePayload, 0);
  const lens: number[] = [];
  for (let i = 0; i < count; i++) {
    const [size, afterSize] = uleb(codePayload, p);
    const start = afterSize;
    let [groups, q] = uleb(codePayload, start);
    for (let g = 0; g < groups; g++) {
      [, q] = uleb(codePayload, q);
      q = skipValtype(codePayload, q);
    }
    lens.push(q - start);
    p = start + size;
  }
  return lens;
}

type Hint = [number, number, string];

function hints(payload: Uint8Array, firstDef: number, lens: number[]): Hint[] {
  let [n, p] = uleb(payload, 0);
  const out: Hint[] = [];
  for (let i = 0; i < n; i++) {
    let f: number, k: number;
    [f, p] = uleb(payload, p);
    [k, p] = uleb(payload, p);
    for (let j = 0; j < k; j++) {
      let off: number, size: number;
      [off, p] = uleb(payload, p);
      [size, p] = uleb(payload, p);
      const val = payload.subarray(p, p + size);
      p += size;
      out.push([f, off - lens[f - firstDef], hex(val)]);
    }
  }
  return out;
}

function importFuncs(secs: Section[]): number | null {
  for (const { id, payload } of secs) {
    if (id !== 2) continue;
    let [n, p] = uleb(payload, 0);
    let funcs = 0;
    for (let i = 0; i < n; i++) {
      for (let k = 0; k < 2; k++) {
        const [ln, q] = uleb(payload, p);
        p = q + ln;
      }
      const kind = payload[p++];
      if (kind === 0) {
        funcs++;
        [, p] = uleb(payload, p);
      } else {
        return null; // other import kinds: not needed by any hint test
      }
    }
    return funcs;
  }
  return 0;
}

const hex = (b: Uint8Array) => Array.from(b, (x) => x.toString(16).padStart(2, "0")).join("");

type Described = { layout: string[]; customs: [string, string | Hint[]][] };

function describe(b: Uint8Array): Described {
  const secs = sections(b);
  const lens = localsLen(secs.find((s) => s.id === 10)?.payload ?? new Uint8Array([0]));
  const first = importFuncs(secs) ?? 0;
  const layout: string[] = [];
  const customs: [string, string | Hint[]][] = [];
  for (const { id, payload } of secs) {
    if (id === 0) {
      const [name, data] = customName(payload);
      layout.push("custom:" + name);
      customs.push(name === "metadata.code.branch_hint"
        ? [name, hints(data, first, lens)]
        : [name, hex(data)]);
    } else {
      layout.push(String(id));
    }
  }
  return { layout, customs };
}

const assemble = (cmd: string, args: string[]) => {
  const r = spawnSync(cmd, args, { encoding: "utf8" });
  return { ok: r.status === 0, err: (r.stderr ?? "").trim() };
};

const wasmrt = resolve(process.argv[2] ?? "");
const files: string[] = [];
// `@list.txt` names a file of paths, one per line — a whole corpus overflows a Windows
// command line.
for (const arg of process.argv.slice(3)) {
  if (arg.startsWith("@")) {
    for (const ln of readFileSync(arg.slice(1), "utf8").split(/\r?\n/)) {
      if (ln.trim()) files.push(ln.trim());
    }
  } else {
    files.push(arg);
  }
}

const wasmTools = spawnSync("wasm-tools", ["--version"], { encoding: "utf8" }).status === 0;
if (!existsSync(wasmrt) || !wasmTools) {
  console.error(`need ${wasmrt} and wasm-tools on PATH`);
  process.exit(2);
}

const tmp = mkdtempSync(join(tmpdir(), "wasmrt-custom-sections-"));
const a = join(tmp, "a.wasm");
const b = join(tmp, "b.wasm");
let bad = 0;
// Agreement that a file is REFUSED compares no bytes at all, so it is counted apart: a probe
// both tools refuse tested nothing, and the headline must not let it pass for a comparison.
const refused: string[] = [];

try {
  for (const f of files) {
    const ra = assemble(wasmrt, ["wat", f, "-o", a]);
    const rb = assemble("wasm-tools", ["parse", f, "-o", b]);
    if (ra.ok !== rb.ok) {
      bad++;
      console.log(`DIFF  ${f}: wasmrt ${ra.ok ? "accepts" : "refuses"}, ` +
        `wasm-tools ${rb.ok ? "accepts" : "refuses"}`);
      console.log(`        wasmrt:     ${ra.err.slice(0, 160)}`);
      console.log(`        wasm-tools: ${rb.err.slice(0, 160)}`);
      continue;
    }
    if (!ra.ok) {
      refused.push(f);
      continue;
    }
    const da = describe(new Uint8Array(readFileSync(a)));
    const db = describe(new Uint8Array(readFileSync(b)));
    if (JSON.stringify(da) !== JSON.stringify(db)) {
      bad++;
      console.log(`DIFF  ${f}`);
      if (JSON.stringify(da.layout) !== JSON.stringify(db.layout)) {
        console.log(`        layout wasmrt:     ${JSON.stringify(da.layout)}`);
        console.log(`        layout wasm-tools: ${JSON.stringify(db.layout)}`);
      }
      for (let i = 0; i < Math.min(da.customs.length, db.customs.length); i++) {
        const [na, xa] = da.customs[i];
        const [nb, xb] = db.customs[i];
        if (na !== nb || JSON.stringify(xa) !== JSON.stringify(xb)) {
          console.log(`        custom '${na}': wasmrt ${JSON.stringify(xa).slice(0, 200)}`);
          console.log(`        custom '${nb}': wasm-tools ${JSON.stringify(xb).slice(0, 200)}`);
        }
      }
    }
  }
} finally {
  rmSync(tmp, { recursive: true, force: true });
}

const compared = files.length - bad - refused.length;
console.log(`custom-sections gate: ${compared} compared byte-for-byte and agree, ` +
  `${refused.length} refused by both, ${bad} differ (of ${files.length})`);
for (const f of refused) console.log(`  refused by both: ${f}`);
process.exit(bad ? 1 : 0);
